### PR TITLE
#3491 Add instance id to sub pipeline blocks

### PR DESCRIPTION
**Pull request by BALEHOK**
## What does this PR do?

- Part of #3491
- Adds support for the nested pipelines to the `withInstanceIds` function

## Discussion

1. Not sure if `withInstanceIds` should validate the pipeline expressions. it's not its responsibility, is it?
2. What's your opinion on `static BLOCK_ID` approach? Should we call `validateRegistryId` or just cast to `RegistryId`?

## Future Work

- Update [omitMetadata](https://github.com/pixiebrix/pixiebrix-extension/blob/4ba9771e3d6d8c3d32557abdc6db3a6bec0a15e9/src/pageEditor/extensionPoints/base.ts#L109-L109) to remove instanceIds from bricks

## Checklist

- [x] Add tests
- [ ] Designate a primary reviewer

### CODE REVIEW

File: src/pageEditor/extensionPoints/withInstanceIds.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            if (draft.config.except != null) {
+              const exceptPipeline = draft.config.except as PipelineExpression;
+              exceptPipeline.__value__ = withInstanceIds(
+                exceptPipeline.__value__
+              );
+            }
+          });
+
+        default:
+          throw "not supported";
```
Comment (twschiller) [inline src/pageEditor/extensionPoints/withInstanceIds.ts:100]: Don't throw strings/literals. Throw an error object

File: src/pageEditor/extensionPoints/withInstanceIds.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+]);
+/**
+ * Enrich a BlockPipeline with instanceIds for use in tracing.
+ */
+
+export function withInstanceIds(blocks: BlockPipeline): BlockPipeline {
+  return blocks.map((blockConfig) => {
+    if (subPipelineBlocks.has(blockConfig.id)) {
+      switch (blockConfig.id) {
+        case ForEach.BLOCK_ID:
```
Comment (twschiller) [inline src/pageEditor/extensionPoints/withInstanceIds.ts:39]: Hard-coding the bricks for now is OK. In the future, we'll need to based this off the inputSchema of the brick (which will require a registry lookup, which introduces some circular dependency trickiness currently)

File: src/pageEditor/extensionPoints/withInstanceIds.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+          if (!isPipelineExpression(blockConfig.config.body)) {
+            throw new Error('ForEach "body" must be a pipeline');
+          }
+
+          return produce(blockConfig, (draft) => {
```
Comment (twschiller) [inline src/pageEditor/extensionPoints/withInstanceIds.ts:44]: All of these branches are very repetitive. Introduce a helper method that takes in the original config and the pipeline property names

```
function withPipelineExpressionInstanceIds(config, pipelineProps) {
   // Check if it's a pipeline expression
   // Call withInstanceIds on the pipeline value
}
```

**Comment by twschiller** [comment]
> Not sure if withInstanceIds should validate the pipeline expressions. it's not its responsibility, is it?

Officially, it's not its responsibility - they get validated at runtime

That being said, I'm in favor of throwing an error here and assuming in the Page Editor that you're dealing with a valid pipeline objects

As part of that, if a branch is left of, I'm OK with this method normalizing the input to have the pipelines as empty arrays so that we can assume a consistent data shape in the Page Editor

> What's your opinion on static BLOCK_ID approach? Should we call validateRegistryId or just cast to RegistryId?

validateRegistryId is the preferred method to avoid any mistakes in writing a registry id

As commented in the PR, I'm OK with the static BLOCK_ID for now. 

In the future (not sure when), we will want to base it on the inputSchema of the brick, as that will allow us to define custom bricks in YAML that take pipelines. (I.e., higher order functions)

File: src/pageEditor/extensionPoints/withInstanceIds.ts (hunk 3 above)
Comment (BALEHOK) [reply src/pageEditor/extensionPoints/withInstanceIds.ts]: Refactored the function. The case when you're happy the tests are there before making changes :)

**Pull request merged**